services/shhext/api.go

```diff
@@ -195,6 +251,125 @@ func (api *PublicAPI) ConfirmMessagesProcessed(messages []*whisper.Message) erro
 	return api.service.deduplicator.AddMessages(messages)
 }
 
+// SendPublicMessage sends a public chat message to the underlying transport
+func (api *PublicAPI) SendPublicMessage(ctx context.Context, msg chat.SendPublicMessageRPC) (hexutil.Bytes, error) {
+	privateKey, err := api.service.w.GetPrivateKey(msg.Sig)
+	if err != nil {
+		return nil, err
+	}
+
+	// This is transport layer agnostic
+	protocolMessage, err := api.service.protocol.BuildPublicMessage(privateKey, msg.Payload)
+	if err != nil {
+		return nil, err
+	}
+
+	symKeyID, err := api.service.w.AddSymKeyFromPassword(msg.Chat)
+	if err != nil {
+		return nil, err
+	}
+
+	// Enrich with transport layer info
+	whisperMessage := chat.PublicMessageToWhisper(&msg, protocolMessage)
+	whisperMessage.SymKeyID = symKeyID
+
+	// And dispatch
+	return api.Post(ctx, *whisperMessage)
+}
+
+// SendDirectMessage sends a 1:1 chat message to the underlying transport
+func (api *PublicAPI) SendDirectMessage(ctx context.Context, msg chat.SendDirectMessageRPC) ([]hexutil.Bytes, error) {
+	if !api.service.pfsEnabled {
+		return nil, ErrPFSNotEnabled
+	}
+	// To be completely agnostic from whisper we should not be using whisper to store the key
+	privateKey, err := api.service.w.GetPrivateKey(msg.Sig)
+	if err != nil {
+		return nil, err
+	}
+
+	publicKey, err := crypto.UnmarshalPubkey(msg.PubKey)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := []*ecdsa.PublicKey{publicKey}
+	// This is transport layer-agnostic
+	protocolMessages, err := api.service.protocol.BuildDirectMessage(privateKey, keys, msg.Payload)
+	if err != nil {
+		return nil, err
+	}
+
+	var response []hexutil.Bytes
+
+	for key, message := range *protocolMessages {
+
+		msg.PubKey = crypto.FromECDSAPub(key)
+		// Enrich with transport layer info
+		whisperMessage := chat.DirectMessageToWhisper(&msg, message)
+
+		// And dispatch
+		hash, err := api.Post(ctx, *whisperMessage)
+		if err != nil {
+			return nil, err
+		}
+		response = append(response, hash)
+
+	}
+	return response, nil
+}
+
+// SendGroupMessage sends a group messag chat message to the underlying transport
+func (api *PublicAPI) SendGroupMessage(ctx context.Context, msg chat.SendGroupMessageRPC) ([]hexutil.Bytes, error) {
+	if !api.service.pfsEnabled {
+		return nil, ErrPFSNotEnabled
+	}
+
+	// To be completely agnostic from whisper we should not be using whisper to store the key
+	privateKey, err := api.service.w.GetPrivateKey(msg.Sig)
+	if err != nil {
+		return nil, err
+	}
+
+	var keys []*ecdsa.PublicKey
```

Comment:
Minor: you can do `keys := make([]*ecdsa.PublicKey, len(msg.PubKeys))` to avoid multiple allocations. Just remember to assign values in `for` loop using index instead of `append()`.